core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala

```diff
@@ -36,7 +38,7 @@ object WeaponRegeXJS {
     *   A JavaScript Array of [[weaponregex.model.mutation.Mutant]] if can be parsed, or throw an exception otherwise
     */
   @JSExportTopLevel("mutate")
-  def mutate(pattern: String, options: MutationOptions = new MutationOptions): js.Array[MutantJS] = {
+  def mutate(pattern: String, flags: String, options: MutationOptions): js.Array[MutantJS] = {
```

Comment:
Overloading doesn't exist in JS. 

Generally, overloading is mimicked with something like this:

```js
function mutate(pattern, flags, options) {
  // fix overloadings. 
  // flags is either a string or options or undefined
  // options is either options or undefined
  if (!options) {
    options = {};
  }
  if (typeof flags === 'string') {
     options.flags = flags;
  } else {
     options = flags;
  }
}
```

I'm not sure if this is supported out-of-the-box by scala-js, maybe @hugo-vrijswijk knows how this works?

Since flags are optional, it might also make sense to put them in the `MutationOptions` class. Although you might not consider it to be a "mutation option".

Reply:
> Since flags are optional, it might also make sense to put them in the `MutationOptions` class. Although you might not consider it to be a "mutation option".

I wouldn't consider it a mutation option so I think that'd be bad API design. I think the current API is the best available.

An alternative would be to make the first argument a class as well (with `pattern` and optional `flags` fields). This might be more sustainable, allowing for further extension of the class in the future. Unfortunately that'd still require method overloading (or a breaking change).

Reply:
I would expect that two exported functions of the same name would create a linking error in Scala.js, but that doesn't seem to be the case. 
Suggestion: either create a `mutateWithFlags`, or have the flags as a _third_ optional argument after options. That wouldn't break anything (at least source compatibility, but js doesn't have binary compatibility 🤠)

Reply:
I did a small test. This:

```scala
 class MutationOptions(
      val mutationLevels: js.Array[Int] = null
  ) extends js.Object
object Parser {
  @JSExportTopLevel("mutate") 
  def mutate(pattern: String, flags: String, options: MutationOptions): Unit = {
  }
  @JSExportTopLevel("mutate") 
  def mutate(pattern: String, options: MutationOptions): Unit = {
  }
}
```

Gets compiled into:

```js
mutate = (function(arg, arg$2, ...rest) {
  switch ($uI(rest.length)) {
    case 0: {
      var prep0 = $as_T(arg);
      var prep1 = arg$2;
      $m_LParser$().mutate__T__LMutationOptions__V(prep0, prep1);
      return (void 0);
      break
    }
    case 1: {
      var prep0$2 = $as_T(arg);
      var prep1$2 = $as_T(arg$2);
      var prep2 = rest[0];
      $m_LParser$().mutate__T__T__LMutationOptions__V(prep0$2, prep1$2, prep2);
      return (void 0);
      break
    }
    default: {
      throw new TypeError("No matching overload")
    }
  }
});
```

So overloads seem to be supported just fine.

I wouldn't worry too much about breaking changes, since we might want to bundle this in a major release anyhow because of https://github.com/stryker-mutator/weapon-regex/issues/166

Reply:
I also had the same doubts before, but Scala.js did some clever work behind the scenes so “overloads” seems to work just fine in JS. Anyway, how do you want the API, since Stryker is virtually the only user of this project 😂

Reply:
In that case, let's do what fits best with your design. Flags as a second argument are fine by me!

Btw, this package has ~50k downloads per week, isn't that amazing!!! 💙